test/PSMVariant1Actions.t.sol

```diff
@@ -0,0 +1,594 @@
+        _deposit(address(this), 100e6);
+        savingsToken.approve(address(actions), 80e18);
+
+        _assertBalances({
+            user:                address(this),
+            gemBalance:          0,
+            daiBalance:          0,
+            savingsTokenBalance: 80e18
+        });
+        _assertZeroBalances(address(receiver));
+        _assertZeroBalances(address(actions));
+
+        uint256 amountOut = actions.redeemAndSwap(receiver, 80e18, 100e6);
+
+        _assertZeroBalances(address(this));
+        _assertBalances({
+            user:                receiver,
+            gemBalance:          100e6,
+            daiBalance:          0,
+            savingsTokenBalance: 0
+        });
+        _assertZeroBalances(address(actions));
+        assertEq(amountOut, 100e6);
+    }
+
+    function test_redeemAndSwap_sameReceiver() public {
+        _deposit(address(this), 100e6);
+        savingsToken.approve(address(actions), 80e18);
+
+        _assertBalances({
+            user:                address(this),
+            gemBalance:          0,
+            daiBalance:          0,
+            savingsTokenBalance: 80e18
+        });
+        _assertZeroBalances(address(actions));
+
+        uint256 amountOut = actions.redeemAndSwap(address(this), 80e18, 100e6);
+
+        _assertBalances({
+            user:                address(this),
+            gemBalance:          100e6,
+            daiBalance:          0,
+            savingsTokenBalance: 0
+        });
+        _assertZeroBalances(address(actions));
+        assertEq(amountOut, 100e6);
+    }
+
+    function test_redeemAndSwap_fee() public {
+        _deposit(address(this), 100.5e6);
+        psm.__setTout(0.005e18);
+        savingsToken.approve(address(actions), 80.4e18);
+
+        _assertBalances({
+            user:                address(this),
+            gemBalance:          0,
+            daiBalance:          0,
+            savingsTokenBalance: 80.4e18
+        });
+        _assertZeroBalances(address(actions));
+
+        uint256 amountOut = actions.redeemAndSwap(address(this), 80.4e18, 100e6);
+
+        _assertBalances({
+            user:                address(this),
+            gemBalance:          100e6,
+            daiBalance:          0,
+            savingsTokenBalance: 0
+        });
+        _assertZeroBalances(address(actions));
+        assertEq(amountOut, 100e6);
+    }
+
+    function testFuzz_redeemAndSwap(
+        uint256 shares,
+        uint256 minAmountOut,
+        uint256 fee
+    ) public {
+        shares = bound(shares, 0, MAX_DAI_AMOUNT);
+        fee    = bound(fee,    0, 1e18);
+
+        uint256 assets = shares * 1.25e18 / 1e18;
+        uint256 expectedAmountOut = assets * 1e18 / (1e12 * (1e18 + fee));
+        minAmountOut = bound(minAmountOut, 0, expectedAmountOut);
+
+        psm.__setTout(fee);
+        savingsToken.approve(address(actions), shares + 1e12);  // Approve some extra for rounding
```

Comment:
+ 1e12 here isn't needed

Reply:
ya good point, don't know why I added that..